Support 1.x versions of the D-Bus client library and switch to env-locale

S-NA opened the pull request:
I am unsure why the upper bound was set to `<1` for `dbus`. The changes to support 1.x were rather trivial and nothing has seemingly broken from what I can tell. I left support for the older versions of the library in case it is important. I am fine with removing older support if that is wanted.

The switch to env-locale from system-locale is my main motivation. As mentioned in the commit, system-locale requires the command `locale` visible in PATH. This is not always the case. See issue #63 for an example.

phuhl commented on src/NotificationCenter/Notifications.hs:
```diff
@@ -392,6 +407,7 @@ notificationDaemon config onNote onCloseNote = do
       "Notify" (onNote (emit client))
     ]
+#endif
 
```
Testing new CI
```suggestion

```

phuhl reviewed (changes requested):
Hi there, 

thanks for contributing :)

It appears, that some of the new dependencies could not be resolved. Take a look in line [383](https://github.com/phuhl/linux_notification_center/pull/121/checks?check_run_id=1694741619#step:6:383) of the output. It appears, that you have to add env-locale to the stack.yaml.

```
Error: While constructing the build plan, the following exceptions were encountered:

In the dependencies for linux-notification-center-1.6.0:
    env-locale needed, but the stack configuration has no specified version  (latest matching
               version is 1.0.0.1)
needed since linux-notification-center is a build target.

Some different approaches to resolving this:

  * Recommended action: try adding the following to your extra-deps
    in /home/runner/work/linux_notification_center/linux_notification_center/stack.yaml:

- env-locale-1.0.0.1@sha256:aedab9b7bbed9f523f9821771b4c67d64f34a00be2be60de1d25b1a97278d475,894

Plan construction failed.
```

S-NA commented:
Hello,

My apologies about that. I normally build without using stack which caused me to overlook changing the dependency in stack.yaml.

phuhl commented:
Hey @S-NA, 

I ran into another issue. It builds fine, as is, but it also does not use dbus-1.2.17 as is, as in the stack.yml file still explicitly dbus-0.10.15 is imported.

Trying to upgrade that to dbus-1.2.17 gave me some issues, as other dependencies started to collide.

My questions here, then:

- What is the main reason to support dbus 1.2, why is it better, is it worth it to make that transition with finding the correct package setup?
  - e.g. one problem is, that we'd need a newer version of xml-conduit, 1.9.0.0 which for some reason is not a preferred version
  - also, get it to build with stack at all is a challenge
- Was the change of dbus motivated/required for the change of env-locale? Is it required?

I need this package, to still build with stack. We can of course leave it as is (because this PR does build) but then, nobody building with stack (the majority I suppose, as the AUR packages are using stack) will have the newer dbus version.

If there are no major benefits from using the newer dbus version, I'd just merge this PR and we have the change to env-locale but not to dbus (due to stack.yml).

What do you think?

phuhl commented:
Just as an info, on how I tried to add dbus, here are my changes to stack.yaml (not giving me a successful build):

```
diff -u --label \#\<buffer\ stack.yaml.\~S-NA-improvements\~\> --label \#\<buffer\ stack.yaml\> /tmp/buffer-content-nVAf5c /tmp/buffer-content-5yT0ht
--- #<buffer stack.yaml.~S-NA-improvements~>
+++ #<buffer stack.yaml>
@@ -41,7 +41,10 @@
 # (e.g., acme-missiles-0.3)
 extra-deps: [ ConfigFile-1.1.4
             , gi-harfbuzz-0.0.3
-            , dbus-0.10.15
+            , dbus-1.2.17
+            , network-3.1.0.1
+            , network-bsd-2.8.1.0
+            , xml-conduit-1.9.0.0
             , env-locale-1.0.0.1
             , gtk3-0.15.4
             , stm-2.5.0.0
@@ -63,6 +66,7 @@
             , gi-graphene-1.0.2
             , ansi-terminal-0.11
             , tagsoup-0.14.7
+            , hslogger-1.3.1.0
             ]
 
 # Override default flag values for local packages and extra-deps

Diff finished.  Sat Jan 16 14:02:57 2021

```

This gives me a strange build error where "MissingH" thinks there is no network-bsd package, even though there should be one. No idea, gave up at that point for now

S-NA commented:
> What is the main reason to support dbus 1.2, why is it better, is it worth it to make that transition with finding the correct package setup?

There is nothing really better about it. The motivation for it is it makes packaging for my distribution easier (NixOS). However, the dbus version does not need to be explicitly pinned to 1.2.17. Supporting versions of the dbus library >= 1 is all that is needed. I went ahead and updated the stack.yaml to remove pinning dbus from 0.10.15 and allow it to be resolved from lts-12.19 which by default provides [dbus-1.0.1](https://www.stackage.org/lts-12.19/package/dbus-1.0.1). (I can pin it explicitly to dbus-1.0.1 in extra-deps if you would like.) I also removed the CPP I added to support pre-1.x versions of the dbus library since you do not seem against letting stack use dbus >= 1 and bumped up the minimum version in cabal file for dbus to reflect this.

> Was the change of dbus motivated/required for the change of env-locale? Is it required?

No, the two changes are not dependent on each other. Sorry about the confusion.

> I need this package, to still build with stack. We can of course leave it as is (because this PR does build) but then, nobody building with stack (the majority I suppose, as the AUR packages are using stack) will have the newer dbus version.
> If there are no major benefits from using the newer dbus version, I'd just merge this PR and we have the change to env-locale but not to dbus (due to stack.yml).
> What do you think?

I understand. Hopefully the current PR's changes are to your liking. While it does not have the package build with dbus-1.2.17, it does allow removing the previously ugly CPP since dbus-1.0.1 supports newer API. There is no urgency to or major benefits from using the latest dbus library version.

Let me know if you want me to clean up the PR by squashing the commits or something like that. As it stands now I am happy to have it merged if you are.

phuhl commented:
> remove pinning dbus from 0.10.15

Stupid me, should have tried that first.

Anyways, looking good, thanks for the effort!
